pkg/clients/aws/eks/eks.go

```diff
@@ -228,6 +270,80 @@ func (e *eksClient) ConnectionToken(name string) (string, error) {
 	return v1Prefix + base64.RawURLEncoding.EncodeToString([]byte(presignedURLString)), nil
 }
 
+// getAMIImage checks to see if the requested image ID is compatible with the
+// AMI images available for the cluster. If no imageID is provided, it picks the most
+// recent available image.
+func (e *eksClient) getAMIImage(requestedImageID, clusterVersion string) (*ec2.Image, error) {
+	images, err := e.getAvailableImages(clusterVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(images) == 0 {
+		return nil, errors.New("No available AMI images were found for the cluster version in this region")
+	}
+
+	if requestedImageID != "" {
+		return getImageWithID(requestedImageID, images)
+	}
+
+	return getMostRecentImage(images), nil
+}
+
+// getAvailableImages retrieves AMI images available for the cluster.
+func (e *eksClient) getAvailableImages(clusterVersion string) ([]*ec2.Image, error) {
+
+	// make sure the provided cluster version is valid
+	r := regexp.MustCompile(`v?(\d+)\.(\d+)`)
+	versionParts := r.FindStringSubmatch(clusterVersion)
+	if len(versionParts) < 3 {
+		return nil, errors.New("Cluster version is empty or invalid")
+	}
+	versionMajor, versionMinor := versionParts[1], versionParts[2]
+
+	// reconstruct the cluster version off of major and minor
+	version := fmt.Sprintf("%v.%v", versionMajor, versionMinor)
+
+	// ami query
+	filters := map[string][]string{
+		"name":  {fmt.Sprintf("*amazon-eks-node-%v*", version)},
+		"state": {"available"},
+	}
+
+	ec2Filters := []*ec2.Filter{}
+	for name, values := range filters {
+		ec2Filters = append(ec2Filters, &ec2.Filter{Name: aws.String(name), Values: aws.StringSlice(values)})
+	}
+
+	out, err := e.amiClient.DescribeImages(
+		&ec2.DescribeImagesInput{
+			Filters: ec2Filters,
+		})
+
+	return out.Images, err
+}
+
+func getMostRecentImage(images []*ec2.Image) *ec2.Image {
+	var result *ec2.Image
+	for _, img := range images {
+		if result == nil || aws.StringValue(result.CreationDate) < aws.StringValue(img.CreationDate) {
+			result = img
```

Comment:
is this an instance of the infamous golang range gotcha? https://github.com/golang/go/wiki/Range#gotchas

i've been bitten by that numerous times and I may have overcorrected to being too paranoid about it now, but hopefully you were able to test and confirm this function correctly finds the most recent image, esp. when the most recent is NOT the last item in the list.

Reply:
good point. using a regular for loop instead